### PR TITLE
Initial plugin

### PR DESCRIPTION
**Pull request by dittonjs**
So this is not finished but is in a working state. It is secure enough that we can deploy it and start testing some of this stuff.

### CODE REVIEW

**Comment by bfcoder** [comment]
Delete the test directory, we can add spec_canvas later as needed

File: app/controllers/proctored_exams_controller.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  def start_quiz
+    require "byebug";
+    headers = {
+      "Content-Type" => "application/json"
+    }
+    plugin = PluginSetting.find_by(name: "quiz_proctor")
+    quiz = HTTParty.get(
+      "#{plugin.settings[:adhesion_url]}/api/proctored_exams?student_id=#{@current_user.id}&proctor_code=#{params[:proctor_code]}&oauth_consumer_key=exams",
```
Comment (bfcoder) [inline app/controllers/proctored_exams_controller.rb]: Use
```ruby
params = {
  student_id: @current_user.id,
  proctor_code: params[:proctor_code],
  oauth_consumer_key: "exams",
}.to_query
```
instead

File: .rubocop.yml [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+Style/LineEndConcatenation:
+  Description: >-
+                 Use \ instead of + or << to concatenate two string literals at
+                 line end.
+  Enabled: false
+
+Metrics/LineLength:
+  Description: 'Limit lines to 200 characters.'
+  StyleGuide: 'https://github.com/bbatsov/ruby-style-guide#80-character-limits'
+  Max: 200
```
Comment (bfcoder) [inline .rubocop.yml]: No

File: app/controllers/proctored_exams_controller.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    # reset_session_for_login
+  end
+
+  def start_quiz
+    headers = {
+      "Content-Type" => "application/json",
+    }
+    plugin = PluginSetting.find_by(name: "quiz_proctor")
+    quiz = HTTParty.get(
+      "#{plugin.settings[:adhesion_url]}/api/proctored_exams?student_id=#{@current_user.id}&proctor_code=#{params[:proctor_code]}&oauth_consumer_key=exams",
```
Comment (bfcoder) [inline app/controllers/proctored_exams_controller.rb]: Use
```ruby
params = {
  student_id: @current_user.id,
  proctor_code: params[:proctor_code],
  oauth_consumer_key: "exams",
}.to_query
```
instead

Comment (bfcoder) [inline app/controllers/proctored_exams_controller.rb]: Don't use a huge string

File: quiz_proctor_engine.gemspec [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+require "quiz_proctor_engine/version"
+
+# Describe your gem and declare its dependencies:
+Gem::Specification.new do |s|
+  s.name        = "quiz_proctor_engine"
+  s.version     = QuizProctorEngine::VERSION
+  s.authors     = ["dittonjs"]
+  s.email       = ["jditton.atomic@gmail.com"]
+  s.homepage    = ""
+  s.summary     = "Summary of QuizProctorEngine."
```
Comment (bfcoder) [inline quiz_proctor_engine.gemspec]: Add a real summary and description

File: app/views/proctored_exams/show.erb [hunk 5]
```diff
@@ -0,0 +1,51 @@
+<style>
```
Comment (bfcoder) [inline app/views/proctored_exams/show.erb]: Why not use a scss file?

File: app/controllers/proctored_exams_controller.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
+require "httparty"
+
+class ProctoredExamsController < ApplicationController
+
+  layout "borderless_lti"
+
+  def show
+    # reset_session_for_login
```
Comment (bfcoder) [inline app/controllers/proctored_exams_controller.rb]: Do you need this?

File: app/controllers/proctored_exams_controller.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    query = {
+      student_id: @current_user.id,
+      proctor_code: params[:proctor_code],
+      oauth_consumer_key: "exams",
+    }.to_query
+
+    quiz = HTTParty.get(
+      "#{plugin.settings[:adhesion_url]}/api/proctored_exams?#{query}",
+      headers: headers,
+      verify: false,
```
Comment (bfcoder) [inline app/controllers/proctored_exams_controller.rb]: I don't think you want this. For dev maybe, but we should be verifying ssl in prod.

File: app/controllers/proctored_exams_controller.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    )
+
+    if quiz.parsed_response["error"].present?
+      flash[:error] = quiz.parsed_response["error"]
+      redirect_to "/proctored_exams"
+    else
+      session[:is_proctored] = true
+      redirect_to "/courses/#{quiz.parsed_response['course_id']}/quizzes/#{quiz.parsed_response['exam_id']}/take"
```
Comment (bfcoder) [inline app/controllers/proctored_exams_controller.rb]: Can you use `course_quiz_take_path` instead?

File: app/views/proctored_exams/show.erb [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    float: right;
+  }
+
+</style>
+<%= render :partial => 'shared/flash_notices' %>
+<div class="jd-main-content">
+  <div class="jd-proctor-code-container">
+    <div class="jd-title">Proctor Code</div>
+    <div>
+      <%= form_tag("/proctored_exams", method: "post") do %>
```
Comment (bfcoder) [inline app/views/proctored_exams/show.erb]: Use `proctored_exams_path`

File: app/controllers/proctored_exams_controller.rb [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+    quiz = HTTParty.get(
+      "#{plugin.settings[:adhesion_url]}/api/proctored_exams?#{query}",
+      headers: headers,
+      verify: false,
+    )
+
+    if quiz.parsed_response["error"].present?
+      flash[:error] = quiz.parsed_response["error"]
+      redirect_to "/proctored_exams"
```
Comment (bfcoder) [inline app/controllers/proctored_exams_controller.rb]: use `proctored_exams_path`

File: app/views/proctored_exams/show.erb [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    background-color: #c1a17a;
+    border: none;
+    border-radius: 3px;
+    color: white;
+    padding: 5px 10px;
+    float: right;
+  }
+
+</style>
+<%= render :partial => 'shared/flash_notices' %>
```
Comment (bfcoder) [inline app/views/proctored_exams/show.erb]: Use the new notation `partial: 'shared/flash_notices'`

File: README.rdoc [hunk 12]
```diff
@@ -0,0 +1,3 @@
+= QuizProctorEngine
```
Comment (jbasdf) [inline README.rdoc]: Add the docs!

File: app/views/proctored_exams/show.erb (hunk 5 above)
Comment (jbasdf) [inline app/views/proctored_exams/show.erb]: This easier. Do what's easier. We're easy!!

File: lib/quiz_proctor_engine/quizzes_controller.rb [hunk 13]
```diff
@@ -0,0 +1,5 @@
+require_dependency "app/controllers/quizzes/quizzes_controller"
```
Comment (jbasdf) [inline lib/quiz_proctor_engine/quizzes_controller.rb]: We can delete this file

File: MIT-LICENSE [hunk 14]
```diff
@@ -0,0 +1,20 @@
+Copyright 2017 dittonjs
```
Comment (fugue88) [inline MIT-LICENSE]: You are not the copyright holder.  This is a work-for-hire.  Atomic Jolt is the copyright holder.

We are supposed to be using the AGPL license whenever possible.  Is there a reason you didn't?

This change needs to be backed out and recommitted correctly.

Comment (dittonjs) [inline MIT-LICENSE]: because all of this got auto generated when we created the engine I never even looked at this file.

Comment (bfcoder) [inline MIT-LICENSE]: Even auto generated files we need to inspect and make sure it is correct. I should have caught this erroneous file. Sorry about that @fugue88.

Comment (dittonjs) [inline MIT-LICENSE]: I understand that it needs to be changed, but the attitude in @fugue88s comment and the email I think was way out of line. We are all on a team here and dont need to be publicly attacked. @fugue88 if you want me to change this you talk to me directly and we can work this out.

Comment (bfcoder) [inline MIT-LICENSE]: There was no attack here @dittonjs. You need not take offense by this.

**Pull request merged**